src/Interpolations.jl

```diff
@@ -123,131 +123,126 @@ function copy_with_padding(A, it::InterpolationType)
-            N->:(error("Linear indexing is not supported for interpolation objects"))
-        ))
-
-        # Resolve ambiguity with real multilinear indexing
-        #   getindex{T,N}(A::AbstractArray{T,N}, NTuple{N,Real}...)
-        eval(ngenerate(:N, :T, :(getindex{T,N,TCoefs}(itp::Interpolation{T,N,TCoefs,$IT,$EB}, x::NTuple{N,Real}...)), getindex_impl))
-
-        # Resolve ambiguity with general array indexing,
-        #   getindex{T,N}(A::AbstractArray{T,N}, I::AbstractArray{T,N})
-        eval(ngenerate(:N, :T, :(getindex{T,N,TCoefs}(itp::Interpolation{T,N,TCoefs,$IT,$EB}, I::AbstractArray{T})),
-            N->:(error("Array indexing is not defined for interpolation objects."))
-        ))
-
-        # Resolve ambiguity with colon indexing for 1D interpolations
-        #   getindex{T}(A::AbstractArray{T,1}, C::Colon)
-        eval(ngenerate(:N, :T, :(getindex{T,TCoefs}(itp::Interpolation{T,1,TCoefs,$IT,$EB}, c::Colon)),
-            N->:(error("Colon indexing is not supported for interpolation objects"))
-        ))
-
-        # Allow multilinear indexing with any types
-        eval(ngenerate(:N, :(promote_type(T,TIndex)), :(getindex{T,N,TCoefs,TIndex}(itp::Interpolation{T,N,TCoefs,$IT,$EB}, x::NTuple{N,TIndex}...)), getindex_impl))
-
-        # Define in-place gradient calculation
-        #   gradient!(g::Vector, itp::Interpolation, NTuple{N}...)
-        eval(ngenerate(:N, :(Vector{TOut}), :(gradient!{T,N,TCoefs,TOut}(g::Vector{TOut}, itp::Interpolation{T,N,TCoefs,$IT,$EB}, x::NTuple{N,Any}...)),
-            N->quote
-                length(g) == $N || error("g must be an array with exactly N elements (length(g) == "*string(length(g))*", N == "*string(N)*")")
-                $(extrap_transform_x(gr,eb,N))
-                $(define_indices(it,N))
-                @nexprs $N dim->begin
-                    @nexprs $N d->begin
-                        (d==dim
-                            ? $(gradient_coefficients(it,N,:d))
-                            : $(coefficients(it,N,:d)))
-                    end
-
-                    @inbounds g[dim] = $(index_gen(degree(it),N))
-                end
-                g
+# Resolve ambiguity with general array indexing,
+#   getindex{T,N}(A::AbstractArray{T,N}, I::AbstractArray{T,N})
+function getindex{T,N,TCoefs,IT,EB}(itp::Interpolation{T,N,TCoefs,IT,EB}, I::AbstractArray{T})
+    error("Array indexing is not defined for interpolation objects.")
+end
+
+# Resolve ambiguity with colon indexing for 1D interpolations
+#   getindex{T}(A::AbstractArray{T,1}, C::Colon)
+function getindex{T,TCoefs,IT,EB}(itp::Interpolation{T,1,TCoefs,IT,EB}, c::Colon)
+    error("Colon indexing is not supported for interpolation objects")
+end
+
+# Allow multilinear indexing with any types
+stagedfunction getindex{T,N,TCoefs,TIndex,IT,EB}(itp::Interpolation{T,N,TCoefs,IT,EB}, x::TIndex...)
+    n = length(x)
+    n == N || return :(error("Must index interpolation objects with as many indexes as dimensions"))
```

Comment:
It would be nice if these error messages also showed `dims(itp)` (i.e. `N`) and `n`, to make it easier to spot bugs (in the client code) where the user has several interpolation objects of different dimensions, and indexes into the wrong one because of a typo.